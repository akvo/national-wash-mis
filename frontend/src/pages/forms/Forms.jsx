import React, { useEffect, useState } from "react";
import { Webform } from "akvo-react-form";
import "akvo-react-form/dist/index.css";
import "./style.scss";
import { useParams, useNavigate } from "react-router-dom";
import { Row, Col, Space, Progress, Result, Button, notification } from "antd";
import { api, store } from "../../lib";
import { takeRight, pick } from "lodash";
import { PageLoader, Breadcrumbs, DescriptionPanel } from "../../components";
import { useNotification } from "../../util/hooks";
import moment from "moment";
import { getTranslation } from "../../util";

const Forms = () => {
  const navigate = useNavigate();
  const { user: authUser } = store.useState((s) => s);
  const { formId } = useParams();
  const [loading, setLoading] = useState(true);
  const [forms, setForms] = useState([]);
  const [percentage, setPercentage] = useState(0);
  const [submit, setSubmit] = useState(false);
  const [showSuccess, setShowSuccess] = useState(false);
  const { notify } = useNotification();
  const { language } = store.useState((s) => s);
  const { active: activeLang } = language;
  const text = getTranslation(activeLang, "forms");

  const formType = window.forms.find(
    (x) => x.id === parseInt(formId)
  )?.type_text;

  const redirectToBatch =
    (formType === "National" && authUser.role.id === 2) ||
    (formType === "County" && authUser.role.id > 2);

  const pagePath = [
    {
      title: text.controlCenter,
      link: "/control-center",
    },
    {
      title:
        authUser?.role?.value === "Data Entry Staff"
          ? authUser.name
          : "Manage Data",
      link:
        authUser?.role?.value === "Data Entry Staff"
          ? "/profile"
          : "/data/manage",
    },
    {
      title: forms.name,
    },
  ];

  const onFinish = (values) => {
    setSubmit(true);
    const questions = forms.question_group
      .map((x) => x.question)
      .flatMap((x) => x);
    const answers = Object.keys(values)
      .map((v) => {
        const question = questions.find((q) => q.id === parseInt(v));
        let val = values[v];
        if (val || val === 0) {
          val =
            question.type === "option"
              ? [val]
              : question.type === "geo"
              ? [val.lat, val.lng]
              : val;
          return {
            question: parseInt(v),
            type: question.type,
            value: val,
            meta: question.meta,
          };
        }
        return false;
      })
      .filter((x) => x);
    const names = answers
      .filter((x) => !["geo", "cascade"].includes(x.type) && x.meta)
      .map((x) => {
        return x.value;
      })
      .flatMap((x) => x)
      .join(" - ");
    const geo = answers.find((x) => x.type === "geo" && x.meta)?.value;
    const administration = answers.find(
      (x) => x.type === "cascade" && x.meta
    )?.value;
    const data = {
      data: {
        administration: administration
          ? takeRight(administration)[0]
          : authUser.administration.id,
        name: names.length
          ? names
          : `${authUser.administration.name} - ${moment().format("MMM YYYY")}`,
        geo: geo || null,
      },
      answer: answers
        .map((x) => {
          if (x.type === "cascade") {
            return { ...x, value: takeRight(x.value)?.[0] || null };
          }
          return x;
        })
        .map((x) => pick(x, ["question", "value"])),
    };
    api
      .post(`form-pending-data/${formId}`, data)
      .then(() => {
        setTimeout(() => {
          setShowSuccess(true);
        }, 3000);
      })
      .catch(() => {
        notification.error({
          message: text.errorSomething,
        });
      })
      .finally(() => {
        setTimeout(() => {
          setSubmit(false);
        }, 2000);
      });
  };

  const onFinishFailed = ({ errorFields }) => {
    if (errorFields.length) {
      notify({
        type: "error",
        message: text.errorMandatoryFields,
      });
    }
  };

  const onChange = ({ progress }) => {
    setPercentage(progress.toFixed(0));
  };

  useEffect(() => {
    if (formId && loading) {
      api.get(`/form/web/${formId}`).then((res) => {
        const questionGroups = res.data.question_group.map((qg) => {
          const questions = qg.question.map((q) => {
            let qVal = { ...q };
            if (q?.extra) {
              delete qVal.extra;
              qVal = {
                ...qVal,
                ...q.extra,
              };
              if (q.extra?.allowOther) {
                qVal = {
                  ...qVal,
                  allowOtherText: text.enterOther,
                };
              }
            }
            return qVal;
          });
          return {
            ...qg,
            question: questions,
          };
        });
        setForms({ ...res.data, question_group: questionGroups });
        setLoading(false);
      });
    }
  }, [formId, loading, text]);

  return (
    <div id="form">
      <Row justify="center" gutter={[16, 16]}>
        <Col span={24} className="webform">
          <Space>
            <Breadcrumbs
              pagePath={pagePath}
              description={
                <p>
                  {text.description1}
                  <br />
                  {text.description2}
                </p>
              }
            />
          </Space>
          <DescriptionPanel
            description={
              <p>
                {text.description1}
                <br />
                {text.description2}
              </p>
            }
          />
          {loading || !formId ? (
            <PageLoader message={text.fetchingForm} />
          ) : (
            !showSuccess && (
              <Webform
                forms={forms}
                onFinish={onFinish}
                onCompleteFailed={onFinishFailed}
                onChange={onChange}
                submitButtonSetting={{ loading: submit }}
                languagesDropdownSetting={{
                  showLanguageDropdown: false,
                  languageDropdownValue: activeLang,
                }}
              />
            )
          )}
          {(!loading || formId) && !showSuccess && (
            <Progress className="progress-bar" percent={percentage} />
          )}
          {!loading && showSuccess && (
            <Result
              status="success"
              title={text?.formSuccessTitle}
              subTitle={
                redirectToBatch
                  ? text?.formSuccessSubTitle
                  : text?.formSuccessSubTitleForAdmin
              }
              extra={[
                <Button
                  type="primary"
                  key="back-button"
                  onClick={() => setShowSuccess(false)}
                >
                  {text.addNew}
                </Button>,
                !redirectToBatch ? (
                  <Button
                    key="manage-button"
                    onClick={() => navigate("/data/manage")}
                  >
                    {text.backManageData}
                  </Button>
                ) : (
                  <Button
                    key="batch-button"
                    onClick={() => navigate("/data/submissions")}
                  >
                    {text.backBatch}
                  </Button>
                ),
              ]}
            />
          )}
        </Col>
      </Row>
    </div>
  );
};

export default Forms;
