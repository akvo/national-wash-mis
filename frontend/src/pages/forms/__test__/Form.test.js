import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";
import axios from "axios";
import Forms from "../Forms";

jest.mock("axios");

describe("pages/forms", () => {
  test("test form when adding new data", async () => {
    const fakeForm = {
      name: "CLTS",
      question_group: [
        {
          name: "Location",
          question: [
            {
              id: 9747540291,
              name: "location",
              order: 1,
              type: "cascade",
              required: true,
              api: {
                endpoint: "/api/v1/administration",
                list: "children",
                initial: 1,
              },
              meta: true,
            },
          ],
        },
        {
          name: "Village Information",
          question: [
            {
              id: 569070281,
              name: "Village",
              order: 1,
              type: "text",
              required: true,
              meta: true,
            },
            {
              id: 567440335,
              name: "Number of households in the Village",
              order: 2,
              type: "number",
              required: true,
              meta: false,
            },
            {
              id: 571070201,
              name: "Take GPS Location of Village",
              order: 3,
              type: "geo",
              required: true,
              center: {
                lat: 9.145,
                lng: 40.4897,
              },
              meta: true,
            },
          ],
        },
        {
          name: "CLTS Intervention Information",
          question: [
            {
              id: 494780324,
              name: "Implementing partner",
              order: 1,
              type: "option",
              required: true,
              option: [
                {
                  id: 10755,
                  name: "Amref",
                  order: 1,
                },
                {
                  id: 10756,
                  name: "Amref and Local Government",
                  order: 2,
                },
                {
                  id: 10757,
                  name: "Other",
                  order: 3,
                },
              ],
              meta: false,
            },
            {
              id: 559830326,
              name: "Date Triggered",
              order: 2,
              type: "date",
              required: false,
              meta: false,
            },
            {
              id: 569070282,
              name: "Number of Latrines at Triggering?",
              order: 3,
              type: "number",
              required: false,
              meta: false,
            },
            {
              id: 569090299,
              name: "Presence of Handwashing Facility with Water and Soap",
              order: 4,
              type: "option",
              required: false,
              option: [
                {
                  id: 10758,
                  name: "Yes",
                  order: 1,
                },
                {
                  id: 10759,
                  name: "No",
                  order: 2,
                },
              ],
              meta: false,
            },
            {
              id: 571070202,
              name: "No Visible Feces were Found in Environment",
              order: 5,
              type: "option",
              required: false,
              option: [
                {
                  id: 10760,
                  name: "Yes",
                  order: 1,
                },
                {
                  id: 10761,
                  name: "No",
                  order: 2,
                },
              ],
              meta: false,
            },
            {
              id: 557710260,
              name: "School Children had Access to Toilets in the Schools",
              order: 6,
              type: "option",
              required: false,
              option: [
                {
                  id: 10762,
                  name: "Yes",
                  order: 1,
                },
                {
                  id: 10763,
                  name: "No",
                  order: 2,
                },
              ],
              meta: false,
            },
            {
              id: 573010345,
              name: "Public Toilets were Available in Public Places",
              order: 7,
              type: "option",
              required: false,
              option: [
                {
                  id: 10764,
                  name: "Yes",
                  order: 1,
                },
                {
                  id: 10765,
                  name: "No",
                  order: 2,
                },
              ],
              meta: false,
            },
            {
              id: 557710261,
              name: "Toilets were Available for Passengers",
              order: 8,
              type: "option",
              required: false,
              option: [
                {
                  id: 10766,
                  name: "Yes",
                  order: 1,
                },
                {
                  id: 10767,
                  name: "No",
                  order: 2,
                },
              ],
              meta: false,
            },
            {
              id: 557690351,
              name: "Date Verified",
              order: 9,
              type: "date",
              required: false,
              meta: false,
            },
            {
              id: 569090301,
              name: "Number of Latrines at Verification Date",
              order: 10,
              type: "number",
              required: false,
              meta: false,
            },
            {
              id: 583770042,
              name: "Latrine Type of Most Latrines at Verification Date",
              order: 11,
              type: "option",
              required: true,
              option: [
                {
                  id: 10768,
                  name: "Bucket",
                  order: 1,
                },
                {
                  id: 10769,
                  name: "Container-based sanitation",
                  order: 2,
                },
                {
                  id: 10770,
                  name: "Flush/pour flush to piped sewer system",
                  order: 3,
                },
                {
                  id: 10771,
                  name: "Flush/pour flush to septic tank",
                  order: 4,
                },
                {
                  id: 10772,
                  name: "Flush/pour flush to pit latrine",
                  order: 5,
                },
                {
                  id: 10773,
                  name: "Flush/pour flush to open drain",
                  order: 6,
                },
                {
                  id: 10774,
                  name: "Hanging toilet/latrine",
                  order: 7,
                },
                {
                  id: 10775,
                  name: "Pit latrine with slab",
                  order: 8,
                },
                {
                  id: 10776,
                  name: "Pit latrine without slab/open pit",
                  order: 9,
                },
                {
                  id: 10777,
                  name: "Twin pit offset latrine",
                  order: 10,
                },
                {
                  id: 10778,
                  name: "Twin pit with latrine slab",
                  order: 11,
                },
                {
                  id: 10779,
                  name: "Twin pit without slab",
                  order: 12,
                },
                {
                  id: 10780,
                  name: "Ventilated improved pit latrine (VIP) with slab",
                  order: 13,
                },
                {
                  id: 10781,
                  name: "Other composting latrine",
                  order: 14,
                },
              ],
              meta: false,
            },
            {
              id: 557700349,
              name: "Open Defecation Status",
              order: 12,
              type: "option",
              required: false,
              option: [
                {
                  id: 10782,
                  name: "Open Defecation",
                  order: 1,
                },
                {
                  id: 10783,
                  name: "Triggered",
                  order: 2,
                },
                {
                  id: 10784,
                  name: "Declared ODF",
                  order: 3,
                },
                {
                  id: 10785,
                  name: "Verified ODF",
                  order: 4,
                },
              ],
              meta: false,
            },
          ],
        },
      ],
    };

    axios.mockResolvedValueOnce({ status: 200, data: fakeForm });
    const { container } = render(<Forms />, { wrapper: MemoryRouter });
    expect(
      screen.getByText(/Please fill up the webform below/i)
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
