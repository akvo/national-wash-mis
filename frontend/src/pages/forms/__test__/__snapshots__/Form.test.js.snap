// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/forms test form when adding new data 1`] = `
<div>
  <div
    id="form"
  >
    <div
      class="ant-row ant-row-center"
      style="margin: -8px -8px -8px -8px;"
    >
      <div
        class="ant-col ant-col-24 webform"
        style="padding: 8px 8px 8px 8px;"
      >
        <div
          class="ant-space ant-space-horizontal ant-space-align-center"
        >
          <div
            class="ant-space-item"
          >
            <div
              class="ant-breadcrumb"
            >
              <span>
                <span
                  class="ant-breadcrumb-link"
                >
                  <a
                    href="/control-center"
                  >
                    <h1
                      class="ant-typography"
                      style="display: inline;"
                    >
                      Control Center
                    </h1>
                  </a>
                </span>
                <span
                  class="ant-breadcrumb-separator"
                >
                  &gt;
                </span>
              </span>
              <span>
                <span
                  class="ant-breadcrumb-link"
                >
                  <a
                    href="/data/manage"
                  >
                    <h2
                      class="ant-typography"
                      style="display: inline;"
                    >
                      Manage Data
                    </h2>
                  </a>
                </span>
                <span
                  class="ant-breadcrumb-separator"
                >
                  &gt;
                </span>
              </span>
              <span>
                <span
                  class="ant-breadcrumb-link"
                >
                  <h3
                    class="ant-typography"
                    style="display: inline;"
                  >
                    
                  </h3>
                </span>
                <span
                  class="ant-breadcrumb-separator"
                >
                  &gt;
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="description-paragraph"
        >
          <p>
            Please fill up the webform below with relevant responses. You will need to answer all mandatory questions before you can submit.
            <br />
            Once you have sumitted a webform, please do not forget to add it as part of a batch and send it for approval.
          </p>
        </div>
        <div
          class="ant-space ant-space-horizontal ant-space-align-start"
          style="width: 100%; min-height: 280px; padding: 30px;"
        >
          <div
            class="ant-space-item"
          >
            <div
              class="ant-space ant-space-horizontal ant-space-align-center"
            >
              <div
                class="ant-space-item"
                style="margin-right: 24px;"
              >
                <div
                  class="ant-spin ant-spin-spinning"
                >
                  <span
                    aria-label="loading"
                    class="anticon anticon-loading anticon-spin ant-spin-dot"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="loading"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
              <div
                class="ant-space-item"
              >
                <h5
                  class="ant-typography"
                  style="margin-bottom: 0px;"
                >
                  Fetching form..
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
